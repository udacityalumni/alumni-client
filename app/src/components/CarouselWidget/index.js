import React, { PropTypes } from 'react';
import List from 'grommet-udacity/components/List';
import ListItem from 'grommet-udacity/components/ListItem';
import FormField from 'grommet-udacity/components/FormField';
import Box from 'grommet-udacity/components/Box';
import Form from 'grommet-udacity/components/Form';
import Button from 'grommet-udacity/components/Button';
import Carousel from 'grommet-udacity/components/Carousel';
import Anchor from 'grommet-udacity/components/Anchor';
import Section from 'grommet-udacity/components/Section';
import Value from 'grommet-udacity/components/Value';
import Heading from 'grommet-udacity/components/Heading';
import EditIcon from 'grommet-udacity/components/icons/base/Edit';
import TrashIcon from 'grommet-udacity/components/icons/base/Trash';
import CheckmarkIcon from 'grommet-udacity/components/icons/base/Checkmark';
import CloseIcon from 'grommet-udacity/components/icons/base/Close';
import AddIcon from 'grommet-udacity/components/icons/base/Add';
import DragIcon from 'grommet-udacity/components/icons/base/Drag';

import styles from './index.module.scss';
import cssModules from 'react-css-modules';

const CarouselWidget = ({
  images,
  onEditImage,
  onAddImage,
  onDeleteImage,
  newImageInput,
  editImageInput,
  currentlyEditing,
  setEditing,
  cancelEditing,
}) => (
  <div className={styles.carouselWidget}>
    <Box className={styles.carouselBox}>
      <Carousel>
        {images.map((image, i) =>
          <Box align="center" justify="center" key={i}>
            <img src={image.url} className={styles.carouselImage} />
            <div className={styles.overlay}>
              <Value
                value={i + 1}
                size="xlarge"
                className={styles.overlayValue}
              />
              <div className={styles.anchorOverlay}>
                <Anchor href={image.url}>
                  {image.url.length > 80 ? `${image.url.slice(0, 80)}...` : image.url}
                </Anchor>
              </div>
            </div>
          </Box>
        )}
      </Carousel>
    </Box>
    <Form className={styles.form}>
      <Box
        direction="row"
        align="center"
        justify="between"
        className={styles.inputBox}
      >
        <FormField>
          <input
            {...newImageInput}
            type="text"
            placeholder="Add an image URL for the carousel"
          />
        </FormField>
        <span>
          <Button
            icon={<AddIcon />}
            onClick={() => onAddImage({ url: newImageInput.value })}
            plain
          />
        </span>
      </Box>
    </Form>
    <Section pad={{ vertical: 'medium' }}>
      <Heading align="center" tag="h3">
        Current Images
      </Heading>
      <List>
        {images.map((image, i) =>
          <ListItem key={i}>
            {currentlyEditing !== null && currentlyEditing === i ?
              <Box
                direction="row"
                align="center"
                justify="between"
                className={styles.inputBox}
              >
                <FormField className={styles.editInputFormField}>
                  <input
                    {...editImageInput}
                    type="text"
                    placeholder={images[currentlyEditing].url}
                  />
                </FormField>
                <span>
                  <Button
                    icon={<CheckmarkIcon />}
                    onClick={() => onEditImage({
                      image: {
                        id: images[currentlyEditing].id,
                        url: editImageInput.value,
                      },
                      index: currentlyEditing,
                    })}
                    plain
                  />
                </span>
                <span>
                  <Button
                    plain
                    icon={<CloseIcon />}
                    onClick={() => cancelEditing(currentlyEditing)}
                  />
                </span>
              </Box>
            :
              <div className={styles.listItemBox}>
                <span className={styles.imageNum}>
                  <Value value={i + 1} size="medium" />
                </span>
                <span style={{ flex: 1 }}>
                  {image.url.length > 80 ? `${image.url.slice(0, 80)}...` : image.url}
                </span>
                <span>
                  <Button
                    plain
                    icon={<EditIcon />}
                    onClick={() => setEditing(i)}
                  />
                </span>
                <span>
                  <Button
                    plain
                    icon={<TrashIcon />}
                    onClick={() => onDeleteImage(i)}
                  />
                </span>
              </div>
            }
          </ListItem>
        )}
      </List>
    </Section>
  </div>
);

CarouselWidget.propTypes = {
  images: PropTypes.array.isRequired,
  onEditImage: PropTypes.func.isRequired,
  onAddImage: PropTypes.func.isRequired,
  onDeleteImage: PropTypes.func.isRequired,
  newImageInput: PropTypes.object.isRequired,
  editImageInput: PropTypes.object.isRequired,
  currentlyEditing: PropTypes.number,
  setEditing: PropTypes.func.isRequired,
  cancelEditing: PropTypes.func.isRequired,
};

export default cssModules(CarouselWidget, styles);
