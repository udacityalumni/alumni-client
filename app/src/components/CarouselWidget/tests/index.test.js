import CarouselWidget from '../index';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import React from 'react';
import data from './mocks';

describe('<CarouselWidget />', () => {
  it('should render with default props', () => {
    const wrapper = shallow(
      <CarouselWidget
        editImageInput={data.editImageInput}
        newImageInput={data.newImageInput}
        images={data.mockedImages}
        onEditImage={e => e}
        onAddImage={e => e}
        onDeleteImage={e => e}
        currentlyEditing={1}
        setEditing={e => e}
        cancelEditing={e => e}
      />
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
