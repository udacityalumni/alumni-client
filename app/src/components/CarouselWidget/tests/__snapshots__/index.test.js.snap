exports[`<CarouselWidget /> should render with default props 1`] = `
<div>
  <Box
    announce={false}
    direction="column"
    focusable={true}
    pad="none"
    responsive={true}
    tag="div">
    <Carousel
      autoplay={true}
      autoplaySpeed={5000}
      infinite={true}
      persistentNav={true}>
      <Box
        align="center"
        announce={false}
        direction="column"
        focusable={true}
        justify="center"
        pad="none"
        responsive={true}
        tag="div">
        <img
          src="http://images.bidnessetc.com/img/960-google-adds-to-talent-bank-with-machine-learning-courses-on-udacity.jpg" />
        <div>
          <Value
            align="center"
            announce={false}
            size="xlarge"
            value={1} />
          <div>
            <Anchor
              href="http://images.bidnessetc.com/img/960-google-adds-to-talent-bank-with-machine-learning-courses-on-udacity.jpg"
              tag="a">
              http://images.bidnessetc.com/img/960-google-adds-to-talent-bank-with-machine-lea...
            </Anchor>
          </div>
        </div>
      </Box>
      <Box
        align="center"
        announce={false}
        direction="column"
        focusable={true}
        justify="center"
        pad="none"
        responsive={true}
        tag="div">
        <img
          src="https://s3-us-west-1.amazonaws.com/udacity-content/images/shared/self-driving-car-engineer-nanodegree.png" />
        <div>
          <Value
            align="center"
            announce={false}
            size="xlarge"
            value={2} />
          <div>
            <Anchor
              href="https://s3-us-west-1.amazonaws.com/udacity-content/images/shared/self-driving-car-engineer-nanodegree.png"
              tag="a">
              https://s3-us-west-1.amazonaws.com/udacity-content/images/shared/self-driving-ca...
            </Anchor>
          </div>
        </div>
      </Box>
    </Carousel>
  </Box>
  <Form
    compact={false}
    fill={false}
    pad="none">
    <Box
      align="center"
      announce={false}
      direction="row"
      focusable={true}
      justify="between"
      pad="none"
      responsive={true}
      tag="div">
      <FormField
        size="medium"
        strong={false}>
        <input
          active={false}
          autofilled={false}
          dirty={false}
          initialValue=""
          invalid={false}
          name="newImageInput"
          placeholder="Add an image URL for the carousel"
          pristine={true}
          touched={true}
          type="text"
          valid={true}
          value=""
          visited={true} />
      </FormField>
      <span>
        <Button
          icon={
            <function Icon() {
                (0, _classCallCheck3.default)(this, Icon);
                return (0, _possibleConstructorReturn3.default)(this, (Icon.__proto__ || (0, _getPrototypeOf2.default)(Icon)).apply(this, arguments));
              }
              a11yTitleId="add-title"
              responsive={true} />
          }
          onClick={[Function onClick]}
          plain={true}
          type="button" />
      </span>
    </Box>
  </Form>
  <Section
    pad={
      Object {
        "vertical": "medium"
      }
    }>
    <Heading
      align="center"
      tag="h3">
      Current Images
    </Heading>
    <List
      role="list">
      <ListItem
        align="center"
        direction="row"
        pad={
          Object {
            "horizontal": "medium",
            "vertical": "small"
          }
        }
        role="listitem"
        separator="bottom">
        <div>
          <span>
            <Value
              align="center"
              announce={false}
              size="medium"
              value={1} />
          </span>
          <span
            style={
              Object {
                "flex": 1
              }
            }>
            http://images.bidnessetc.com/img/960-google-adds-to-talent-bank-with-machine-lea...
          </span>
          <span>
            <Button
              icon={
                <function Icon() {
                    (0, _classCallCheck3.default)(this, Icon);
                    return (0, _possibleConstructorReturn3.default)(this, (Icon.__proto__ || (0, _getPrototypeOf2.default)(Icon)).apply(this, arguments));
                  }
                  a11yTitleId="edit-title"
                  responsive={true} />
              }
              onClick={[Function onClick]}
              plain={true}
              type="button" />
          </span>
          <span>
            <Button
              icon={
                <function Icon() {
                    (0, _classCallCheck3.default)(this, Icon);
                    return (0, _possibleConstructorReturn3.default)(this, (Icon.__proto__ || (0, _getPrototypeOf2.default)(Icon)).apply(this, arguments));
                  }
                  a11yTitleId="trash-title"
                  responsive={true} />
              }
              onClick={[Function onClick]}
              plain={true}
              type="button" />
          </span>
        </div>
      </ListItem>
      <ListItem
        align="center"
        direction="row"
        pad={
          Object {
            "horizontal": "medium",
            "vertical": "small"
          }
        }
        role="listitem"
        separator="bottom">
        <Box
          align="center"
          announce={false}
          direction="row"
          focusable={true}
          justify="between"
          pad="none"
          responsive={true}
          tag="div">
          <FormField
            size="medium"
            strong={false}>
            <input
              active={false}
              autofilled={false}
              dirty={false}
              initialValue=""
              invalid={false}
              name="editImageInput"
              placeholder="https://s3-us-west-1.amazonaws.com/udacity-content/images/shared/self-driving-car-engineer-nanodegree.png"
              pristine={true}
              touched={true}
              type="text"
              valid={true}
              value=""
              visited={true} />
          </FormField>
          <span>
            <Button
              icon={
                <function Icon() {
                    (0, _classCallCheck3.default)(this, Icon);
                    return (0, _possibleConstructorReturn3.default)(this, (Icon.__proto__ || (0, _getPrototypeOf2.default)(Icon)).apply(this, arguments));
                  }
                  a11yTitleId="checkmark-title"
                  responsive={true} />
              }
              onClick={[Function onClick]}
              plain={true}
              type="button" />
          </span>
          <span>
            <Button
              icon={
                <function Icon() {
                    (0, _classCallCheck3.default)(this, Icon);
                    return (0, _possibleConstructorReturn3.default)(this, (Icon.__proto__ || (0, _getPrototypeOf2.default)(Icon)).apply(this, arguments));
                  }
                  a11yTitleId="close-title"
                  responsive={true} />
              }
              onClick={[Function onClick]}
              plain={true}
              type="button" />
          </span>
        </Box>
      </ListItem>
    </List>
  </Section>
</div>
`;
